File: sdks/java/src/main/java/io/octostore/exceptions/Exceptions.java
package io.octostore.exceptions;

/**
 * Base exception for all OctoStore errors.
 */
class OctoStoreException extends Exception {
    public OctoStoreException(String message) {
        super(message);
    }
    
    public OctoStoreException(String message, Throwable cause) {
        super(message, cause);
    }
}

/**
 * Authentication failed exception.
 */
class AuthenticationException extends OctoStoreException {
    public AuthenticationException(String message) {
        super(message);
    }
}

/**
 * Network error exception.
 */
class NetworkException extends OctoStoreException {
    public NetworkException(String message, Throwable cause) {
        super(message, cause);
    }
}

/**
 * Lock operation failed exception.
 */
class LockException extends OctoStoreException {
    private final int statusCode;
    
    public LockException(String message, int statusCode) {
        super(message);
        this.statusCode = statusCode;
    }
    
    public int getStatusCode() {
        return statusCode;
    }
}

/**
 * Lock is currently held exception.
 */
class LockHeldException extends OctoStoreException {
    private final String holderId;
    private final String expiresAt;
    
    public LockHeldException(String message, String holderId, String expiresAt) {
        super(message);
        this.holderId = holderId;
        this.expiresAt = expiresAt;
    }
    
    public String getHolderId() {
        return holderId;
    }
    
    public String getExpiresAt() {
        return expiresAt;
    }
}

/**
 * Input validation failed exception.
 */
class ValidationException extends RuntimeException {
    public ValidationException(String message) {
        super(message);
    }
}

File: sdks/python/octostore/exceptions.py
"""OctoStore exception classes."""

class OctoStoreError(Exception):
    """Base exception for all OctoStore errors."""
    pass


class AuthenticationError(OctoStoreError):
    """Raised when authentication fails."""
    pass


class NetworkError(OctoStoreError):
    """Raised when network operations fail."""
    pass


class LockError(OctoStoreError):
    """Raised when lock operations fail."""
    pass


class LockHeldError(LockError):
    """Raised when trying to acquire a lock that's already held."""
    
    def __init__(self, message: str, holder_id: str = None, expires_at: str = None):
        super().__init__(message)
        self.holder_id = holder_id
        self.expires_at = expires_at


class ValidationError(OctoStoreError):
    """Raised when input validation fails."""
    pass